src/QsCompiler/Transformations/CallGraph/CallGraphWalkerBase.cs

```diff
@@ -0,0 +1,87 @@
+            where TGraph : CallGraphBuilder<TNode, TEdge>
+            where TNode : CallGraphNodeBase
+            where TEdge : CallGraphEdgeBase
+        {
+            public abstract class TransformationState
+            {
+                internal TNode CurrentNode;
+                internal readonly TGraph Graph;
+
+                // The type parameter resolutions of the current expression.
+                internal IEnumerable<TypeParameterResolutions> ExprTypeParamResolutions = new List<TypeParameterResolutions>();
+                internal QsNullable<Position> CurrentStatementOffset;
+                internal QsNullable<Range> CurrentExpressionRange;
+                internal readonly Stack<TNode> RequestStack = new Stack<TNode>(); // Used to keep track of the nodes that still need to be walked by the walker.
+                internal readonly HashSet<TNode> ResolvedNodeSet = new HashSet<TNode>(); // Used to keep track of the nodes that have already been walked by the walker.
+
+                internal TransformationState(TGraph graph)
+                {
+                    this.Graph = graph;
+                }
+
+                /// <summary>
+                /// Adds dependency to the graph from the current callable to the callable referenced by the given identifier.
+                /// </summary>
+                internal abstract void AddDependency(QsQualifiedName identifier);
+            }
+
+            public class StatementWalker<TState> : StatementTransformation<TState>
+                where TState : TransformationState
+            {
+                public StatementWalker(SyntaxTreeTransformation<TState> parent) : base(parent, TransformationOptions.NoRebuild)
+                {
+                }
+
+                public override QsStatement OnStatement(QsStatement stm)
+                {
+                    this.SharedState.CurrentStatementOffset = stm.Location.IsValue
+                        ? QsNullable<Position>.NewValue(stm.Location.Item.Offset)
+                        : QsNullable<Position>.Null;
+                    return base.OnStatement(stm);
+                }
+            }
+
+            public class ExpressionWalker<TState> : ExpressionTransformation<TState>
+                where TState : TransformationState
+            {
+                public ExpressionWalker(SyntaxTreeTransformation<TState> parent) : base(parent, TransformationOptions.NoRebuild)
+                {
+                }
+
+                public override TypedExpression OnTypedExpression(TypedExpression ex)
+                {
+                    var contextRange = this.SharedState.CurrentExpressionRange;
+                    this.SharedState.CurrentExpressionRange = ex.Range;
+
+                    if (ex.TypeParameterResolutions.Any())
+                    {
+                        this.SharedState.ExprTypeParamResolutions = this.SharedState.ExprTypeParamResolutions.Prepend(ex.TypeParameterResolutions);
```

Comment:
I might be misunderstanding this, but shouldn't the resolutions be removed from the list after this expression is processed, similar to how `CurrentExpressionRange` is reset, otherwise the list could carry over resolutions from unrelated expressions?

Reply:
The ExprTypeParamResolutions are supposed to accumulate resolutions from the top of a nested expression hierarchy to the bottom where an identifier will be found. At the identifier, the list is cleared after it is used. See the AddDependencies method of the other walker classes.

Reply:
Could you give an example of an expression where resolutions propagate down to a (global) identifier? I am wondering what happens if there is no identifier at the bottom (do the resolutions not get cleared?), or if there is more than one identifier adjacent to each other in the same expression.

Reply:
Any expression where not all of the generic type parameters are not resolved on the identifier directly would provide an example of this. The syntax tree would be considered malformed and would have thrown an error before this if there were type parameter resolutions in an expression did not have an identifier at one of the leaves. If there are multiple identifiers in the expression hierarchy, the resolutions for the a given identifier would not be encountered until all the identifiers to the left are already processed, which is the order that the expression hierarchy is parsed by the transformation.

Reply:
So your typical generic call:
```
Foo<'A>(a : 'A) : Unit { }
...
Bar() : Unit {
    Foo(12.0);
}
```
The call to Foo needs to carry the type parameter resolution from the call-like expression `Foo(12.0)` to the identifier sub-expression `Foo`.

An example with multiple parameters:
```
Foo<'A, 'B>(a : 'A, b : 'B) : Unit { }
...
Bar() : Unit {
    Foo<Int, _>(3, 12.0);
}
```
Here the B parameter is resolved to Double by the call-like expression `Foo<Int, _>(3, 12.0)` while the A parameter is resolved to Int by the identifier sub-expression `Foo<Int, _>`.